pallets/providers/src/tests.rs

```diff
@@ -4808,166 +4814,426 @@ mod slash {
-                    bob_deposit_on_hold.saturating_sub(bob_slash_amount)
+                let capacity_deficit =
+                    pre_state_provider.capacity_used - pre_state_provider.capacity;
+
+                let expected_delta_hold_amount =
+                    StorageDataUnitAndBalanceConverter::convert(capacity_deficit);
+
+                let existential_deposit = ExistentialDeposit::get();
+                // Set provider's balance to cover the expected_delta_hold_amount
+                NativeBalance::set_balance(
+                    &self.account,
+                    expected_delta_hold_amount + existential_deposit,
                 );
 
-                // If slash amount is greater than deposit then the actual slash amount should be the deposit amount
-                let actual_alice_slashed_amount = alice_slash_amount.min(alice_deposit_on_hold);
-                let actual_bob_slashed_amount = bob_slash_amount.min(bob_deposit_on_hold);
+                // Top up the provider
+                assert_ok!(StorageProviders::top_up_deposit(RuntimeOrigin::signed(
+                    self.account
+                ),));
 
-                // Check that the Treasury has received the slash amount
+                // Check event TopUpFulfilled is emitted
+                System::assert_has_event(
+                    Event::<Test>::TopUpFulfilled {
+                        provider_id: self.provider_id,
+                        amount: expected_delta_hold_amount,
+                    }
+                    .into(),
+                );
+
+                // Check that the held deposit covers the used capacity
                 assert_eq!(
-                    NativeBalance::free_balance(&<Test as crate::Config>::Treasury::get()),
-                    treasury_balance + actual_alice_slashed_amount + actual_bob_slashed_amount
+                    NativeBalance::balance_on_hold(
+                        &StorageProvidersHoldReason::get(),
+                        &self.account
+                    ),
+                    pre_state_held_amount + expected_delta_hold_amount
                 );
+
+                // Check that the storage has been cleared
+                assert!(GracePeriodToSlashedProviders::<Test>::get(
+                    end_block_grace_period,
+                    self.provider_id
+                )
+                .is_none());
+                assert!(AwaitingTopUpFromProviders::<Test>::get(self.provider_id).is_none());
+            }
+        }
+
+        #[test]
+        fn slash_storage_provider_automatic_top_up() {
+            ExtBuilder::build().execute_with(|| {
+                let mut test_setup = TestSetup::default();
+
+                test_setup.slash_held_deposit = true;
+                test_setup.slash_and_verify();
+
+                test_setup.automatic_top_up = true;
+                test_setup.slash_and_verify();
+            });
+        }
```

Comment:
Took me more time than I'm willing to admit, but... I understood why any of your tests fail if the BSP can top up immediately after having a capacity deficit. None of your tests have `automatic_top_up` enabled the first time you `slash_and_verify`. If you do that, slashing fails.

Reply:
yes exactly. I added this test to cover this case :)